File: LeetCodeSolutions/LeetCode_0990.py
class DSU:
    def __init__(self):
        self.parent = [_ for _ in range(26)]

    def find(self, s):
        s_id = ord(s) - ord('a')
        while self.parent[s_id] != s_id:
            self.parent[s_id] = self.parent[self.parent[s_id]]
            s_id = self.parent[s_id]
        return s_id

    def union(self, a, b):
        root_a = self.find(a)
        root_b = self.find(b)
        if root_a == root_b:
            return
        else:
            self.parent[root_a] = root_b


class Solution:
    def equationsPossible(self, equations: List[str]) -> bool:
        equals = [[_[0], _[3]] for _ in equations if _[1] == '=']
        nequals = [[_[0], _[3]] for _ in equations if _[1] == '!']
        dsu = DSU()
        for a, b in equals:
            dsu.union(a, b)

        for a, b in nequals:
            if dsu.find(a) == dsu.find(b):
                return False
        return True

File: LeetCodeSolutions/LeetCode_0990.cpp
class DSU{
    public:
    unordered_map<int, int> parent;

    int find(int i){
        if (!parent.count(i)) parent[i] = i;
        while (parent[i] != i){
            parent[i] = parent[parent[i]];
            i = parent[i];
        }
        return i;
    }

    void union_(int a, int b){
        int a_root = find(a);
        int b_root = find(b);
        if (a_root != b_root){
            parent[a_root] = b_root;
        }
    }

    bool isEqual(int a, int b){
        return find(a) == find(b);
    }

};

class Solution {
public:
    bool equationsPossible(vector<string>& equations) {
        DSU dsu;
        vector<pair<int, int> > nequals;
        for (string &s: equations){
            if (s[1] == '='){
                dsu.union_(int(s[0] - 'a'), int(s[3] - 'a'));
            }
            else nequals.push_back(make_pair(s[0] - 'a', s[3] - 'a'));
        }
        for (auto &p: nequals){
            if (dsu.isEqual(p.first, p.second)) return false;
        }
        return true;
    }
};